src/nbl/asset/utils/CHLSLCompiler.cpp

```diff
@@ -0,0 +1,332 @@
+            { "vertex", IShader::ESS_VERTEX },
+            { "fragment", IShader::ESS_FRAGMENT },
+            { "tesscontrol", IShader::ESS_TESSELLATION_CONTROL },
+            { "tesseval", IShader::ESS_TESSELLATION_EVALUATION },
+            { "geometry", IShader::ESS_GEOMETRY },
+            { "compute", IShader::ESS_COMPUTE }
+        };
+
+        auto found = stageFromIdent.find(shaderStageIdentifier);
+        if (found == stageFromIdent.end())
+        {
+            return std::string("#error Malformed shader_stage pragma, unknown stage");
+        }
+        stage = found->second;
+
+        if (!lexer.HasNextToken()) return std::string("#error Malformed shader_stage pragma");
+        token = lexer.GetNextToken();
+        if (token.mType != tcpp::E_TOKEN_TYPE::CLOSE_BRACKET) return std::string("#error Malformed shader_stage pragma");
+
+        while (lexer.HasNextToken()) {
+            auto token = lexer.GetNextToken();
+            if (token.mType == tcpp::E_TOKEN_TYPE::NEWLINE) break;
+            if (token.mType != tcpp::E_TOKEN_TYPE::SPACE) return std::string("#error Malformed shader_stage pragma");
+        }
+
+        return std::string("");
+    });
+
+    auto resolvedString = proc.Process();
+    IShaderCompiler::reenableDirectives(resolvedString);
+    return resolvedString;
+}
+
+core::smart_refctd_ptr<ICPUShader> CHLSLCompiler::compileToSPIRV(const char* code, const IShaderCompiler::SCompilerOptions& options) const
+{
+    auto hlslOptions = option_cast(options);
+
+    if (!code)
+    {
+        hlslOptions.preprocessorOptions.logger.log("code is nullptr", system::ILogger::ELL_ERROR);
+        return nullptr;
+    }
+
+    auto stage = hlslOptions.stage;
+    auto newCode = preprocessShader(code, stage, hlslOptions.preprocessorOptions);
+
+    // Suffix is the shader model version
+    std::wstring targetProfile(L"XX_6_2");
+
+    // Set profile two letter prefix based on stage
+    switch (stage) {
+    case asset::IShader::ESS_VERTEX:
+        targetProfile.replace(0, 2, L"vs");
+        break;
+    case asset::IShader::ESS_TESSELLATION_CONTROL:
+        targetProfile.replace(0, 2, L"ds");
+        break;
+    case asset::IShader::ESS_TESSELLATION_EVALUATION:
+        targetProfile.replace(0, 2, L"hs");
+        break;
+    case asset::IShader::ESS_GEOMETRY:
+        targetProfile.replace(0, 2, L"gs");
+        break;
+    case asset::IShader::ESS_FRAGMENT:
+        targetProfile.replace(0, 2, L"ps");
+        break;
+    case asset::IShader::ESS_COMPUTE:
+        targetProfile.replace(0, 2, L"cs");
+        break;
+    case asset::IShader::ESS_TASK:
+        targetProfile.replace(0, 2, L"as");
+        break;
+    case asset::IShader::ESS_MESH:
+        targetProfile.replace(0, 2, L"ms");
+        break;
+    default:
+        hlslOptions.preprocessorOptions.logger.log("invalid shader stage %i", system::ILogger::ELL_ERROR, stage);
+        return nullptr;
+    };
+
+    LPCWSTR arguments[] = {
```

Comment:
is that really all there is to handle in `hlslOptions` ? what about optimization levels, etc? @deprilula28